File: go/pkg/encoders/bytes.go
package encoders

import (
	"github.com/splunk/stef/go/pkg"
)

type BytesEncoder struct {
	StringEncoder
}

type BytesEncoderDict StringEncoderDict

func (e *BytesEncoderDict) Init(limiter *pkg.SizeLimiter) {
	(*StringEncoderDict)(e).Init(limiter)
}

func (e *BytesEncoderDict) Reset() {
	(*StringEncoderDict)(e).Reset()
}

func (e *BytesEncoder) Init(dict *BytesEncoderDict, limiter *pkg.SizeLimiter, columns *pkg.WriteColumnSet) error {
	return e.StringEncoder.Init((*StringEncoderDict)(dict), limiter, columns)
}

func (e *BytesEncoder) Encode(val pkg.Bytes) {
	e.StringEncoder.Encode(string(val))
}

type BytesDecoder struct {
	StringDecoder
}

type BytesDecoderDict StringDecoderDict

func (d *BytesDecoderDict) Init() {
	(*StringDecoderDict)(d).Init()
}

func (d *BytesDecoder) Continue() {
	d.StringDecoder.Continue()
}

func (d *BytesDecoder) Decode(dst *pkg.Bytes) error {
	return d.StringDecoder.Decode((*string)(dst))
}

func (d *BytesDecoder) Init(dict *BytesDecoderDict, columns *pkg.ReadColumnSet) error {
	return d.StringDecoder.Init((*StringDecoderDict)(dict), columns)
}

File: rust/stef-core/src/codecs/bytes.rs
use crate::{codecs::string::{StringDecoder, StringEncoder}, types::Bytes, recordbuf::{ReadColumnSet, WriteColumnSet}, SizeLimiter};

/// Bytes encoder using same layout as string codec.
#[derive(Default)]
pub struct BytesEncoder {
    inner: StringEncoder,
}

impl BytesEncoder {
    pub fn init(&mut self, limiter: &mut SizeLimiter, columns: &mut WriteColumnSet) {
        self.inner.init(limiter, columns);
    }

    pub fn encode(&mut self, val: &[u8]) {
        let s = String::from_utf8_lossy(val).to_string();
        self.inner.encode(&s);
    }

    pub fn collect_columns(&mut self, column_set: &mut WriteColumnSet) {
        self.inner.collect_columns(column_set);
    }

    pub fn reset(&mut self) {
        self.inner.reset();
    }
}

/// Bytes decoder.
#[derive(Default)]
pub struct BytesDecoder {
    inner: StringDecoder,
}

impl BytesDecoder {
    pub fn init(&mut self, columns: &mut ReadColumnSet) {
        self.inner.init(columns);
    }

    pub fn continue_(&mut self) {
        self.inner.continue_();
    }

    pub fn decode(&mut self, dst: &mut Bytes) -> crate::errors::Result<()> {
        let mut s = String::new();
        self.inner.decode(&mut s)?;
        *dst = s.into_bytes();
        Ok(())
    }

    pub fn reset(&mut self) {
        self.inner.reset();
    }
}
